miden-lib/asm/kernels/transaction/api.masm

```diff
@@ -510,23 +510,49 @@ export.get_output_notes_hash
     # => [COM]
 end
 
-#! Creates a new note and returns a pointer to the memory address at which the note is stored.
+#! Creates a new note and returns a pointer to the memory address 
+#! at which the note is stored.
 #!
-#! Inputs: [ASSET, tag, note_type, RECIPIENT]
-#! Outputs: [ptr, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+#! Inputs: [tag, aux, note_type, RECIPIENT]
+#! Outputs: [ptr, 0, 0, 0, 0, 0, 0]
 #!
-#! ASSET is the asset to be included in the note.
 #! tag is the tag to be included in the note.
-#! note_type is the note storage type
+#! aux is an auxiliary field that the user can use.
+#! note_type is the note storage type.
 #! RECIPIENT is the recipient of the note.
 #! ptr is the pointer to the memory address at which the note is stored.
 export.create_note
     # authenticate that the procedure invocation originates from the account context
     exec.authenticate_account_origin
-    # => [ASSET, tag, note_type, RECIPIENT]
+    # => [tag, aux, note_type, RECIPIENT]
 
     exec.tx::create_note
-    # => [ptr, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+    # => [ptr]
+
+    # prepare the stack for return
+    #padw push.0 push.0 movup.6 
+    # => [ptr, 0, 0, 0, 0, 0, 0]
```

Comment:
Got it, so we want to prevent that `x, X, X` moves?

That would be prevented by adding a `padw push.0 push.0 movup.6` at the end. it adds exactly `0, 0, ZERO` in the middle of the 16 elements. 

But I get

```
thread 'tests::test_tx::test_create_note_without_asset' panicked at miden-lib/src/tests/test_tx.rs:63:40:
called `Result::unwrap()` on an `Err` value: InvalidStackDepthOnReturn(22)
```

Reply:
Thanks for the in-depth explanation. 

I changed the `asm/api.masm` `create_note` procedure now to 

```
export.create_note
    # authenticate that the procedure invocation originates from the account context
    exec.authenticate_account_origin
    # => [tag, aux, note_type, RECIPIENT]

    exec.tx::create_note
    # => [ptr]

    # prepare the stack for return
    movupw.3 movup.15 movup.15 movup.6 
    # => [ptr, 0, 0, 0, 0, 0, 0]
end
```

This should do the job. 

When entering the procedure, we have 

```
[tag, aux, note_type, RECIPIENT, x, X, X]
```

after `exec.tx::create_note` we should have (without padding)

```
[ptr, x, X, X, 0, 0, ZERO]
```

so we can do 

```
movupw.3 movup.15 movup.15 movup.6 
```

to reach 

```
[ptr, 0, 0, ZERO, x, X, X]
```

Reply:
I think we should document the changes to the stack in the source code too. It really helps to understand the instructions `movupw.3 movup.15 movup.15 movup.6`.

```
# => [ptr, x, X, X, 0, 0, ZERO]

movupw.3
# => [ZERO, ptr, x, X, X, 0, 0]

movup.15 movup.15
# => [0, 0, ZERO, ptr, x, X, X]

movup.6 
# => [ptr, 0, 0, ZERO, x, X, X]
```